_format/1.0/index.md

```diff
@@ -1104,13 +1114,16 @@ remain consistent with JSON:API's [sorting rules](#fetching-sorting).
 The `page` query parameter is reserved for pagination. Servers and clients
 **SHOULD** use this key for pagination operations.
 
-> Note: JSON:API is agnostic about the pagination strategy used by a server.
-Effective pagination strategies include (but are not limited to):
-page-based, offset-based, and cursor-based. The `page` query parameter can
+Ideally, implementations **SHOULD** use cursor-based strategy for pagination. The 
```

Comment:
Wouldn't it be more effective to only support one type of pagination strategy? Seems like cursor pagination would be great for smaller data sets too.

Reply:
Ideally yes. However, I think the reality is that cursor pagination can be really challenging to implement sometimes. So I would prefer not to make it a requirement.

Reply:
I know I'm gonna sound like a broken record here, but client side will always prefer a consistent good over an inconsistent great.

Reply:
And, for what it's worth, integrating with a page based or offset as opposed to cursor is far simpler for a UI to consume.

I could be convinced, but I'm not entirely sure what the benefit is of cursor based pagination given our use primary use cases.

Reply:
The problem with page or offset based strategies is it limits the choice of datastore on the backend.  Many NoSQL databases either can't even properly handle page-based queries, doing such queries is wildly inefficient.

Using `self`, `next`, and `previous` links is an abstraction that will work regardless of choice of datastore, so it would give us the most consistency.  As we've mentioned that comes at the cost of being more challenging to implement, and somewhat less flexible in the UI if we need jump-to-page functionality.

Reply:
If that's the consensus then that's ok. My goal is to avoid different pagination strategies among similar types of data across the application. 

I recognize that's less of a spec than it is a best practice that requires specific types of data we use, so if we're allowing for different types of pagination it might not be able to be hashed out here.

Reply:
This has been a great discussion and has helped me clarify in my mind what I intended to say. So let me try this again (and if this sounds good, I can update the document). There are two parts to the story here:

* The contract to the client. This MUST always be cursor based. That is, the client will always expect the response to include links for `next` and `previous` as mentioned by Nick. Those links will be cursor based as specified here: https://jsonapi.org/profiles/ethanresnick/cursor-pagination/. So the client should NEVER try and construct those URLs itself -that should be opaque. However the server constructs those URLs, the cursor tokens, is of no consequence to the client. So @jsklarsky-w2g  that would mean  from a client perspective, it would be completely consistent interface and contract. 

* The other part is the actual server implementation of that cursor based scheme. Ideally, the server would try and use cursors (postgres supports this). But if not, it should use time based offsets, but that would NOT be exposed to the client. The "cursor" tokens would still be opaque (it could be the timestamp base64 encoded for example). So again, the client has no knowledge of the actual "cursor" based implementation. It gets the same interface always.

Does that make sense?

Reply:
100%.

Reply:
@jsklarsky-w2g does that also mean you are on board with that approach? :)

Reply:
Good explanation.

Reply:
So to summarize, the client just has to construct the first request, and all subsequent relative requests are provided as links from the server. We are good with this approach.

Reply:
that is correct. Generate the first fetch/search request. The response will include pagination links that should be followed by the client.

Reply:
And in my experience, constructing the first request can completely omit any paging parameters and allow the server to infer a sensible default (usually the first page).

Reply:
But the user could specify a page size if they wanted right? (though if we move to infinite scroll this may become irrelevant on the front end though)

Reply:
Yea, I think the client specifying a page size in the first request is reasonable and common. Perhaps the document should indicate these should be optional in the first request and the server should default to something sensible.